akd/src/directory.rs

```diff
@@ -459,7 +459,15 @@ where
 
         let current_azks = self.retrieve_azks().await?;
         let current_epoch = current_azks.get_latest_epoch();
-        let mut user_data = self.storage.get_user_data(akd_label).await?.states;
+        // Ignore states in storage which are ahead of the current directory epoch
+        let mut user_data: Vec<ValueState> = self
+            .storage
+            .get_user_data(akd_label)
+            .await?
+            .states
+            .into_iter()
+            .filter(|vs| vs.epoch <= current_epoch)
+            .collect();
```

Comment:
Suggestion: Use `retain` and make the filtering a one-liner, like we do with sorting?

Reply:
We can! The code becomes something like this:

```rust
        let mut user_data = self
            .storage
            .get_user_data(akd_label)
            .await?
            .states;

        // Ignore states in storage which are ahead of the current directory epoch
        user_data.retain(|vs| vs.epoch <= current_epoch);
        // reverse sort from highest epoch to lowest
        user_data.sort_by(|a, b| b.epoch.cmp(&a.epoch));
```

Without benchmarking, I think perf should be comparable between the two. That said, `retain` used to be slower. However, it looks like that was fixed: https://github.com/rust-lang/rust/issues/91497.